### PR TITLE
Add snapshot testing for API clients and use them as mock for E2E tests

### LINKED ISSUE #643: [CI] Refactor recipe-on-staging
Currently we use e2e tests on staging : https://github.com/etalab/annuaire-entreprises-site/actions/workflows/recipe-staging.yml

Since we protect staging with http basic auth and current strategy is not pretty bad as we test both API and staging site, we muste move to a new test strategy that only tests API responses.

Current responses mock are stored here :
https://github.com/etalab/annuaire-entreprises-site/tree/main/clients-mocks

### PR DESCRIPTION
**Pull request by johangirod**
- Add a new command `test:client-api` to run snapshot tests for API
  clients
- Remove old mocks
- Use snapshots for E2E testing
- Warns about clients calls that are not mocked when launching E2E tests
- [x] Adds a github action that launches test on PR and every day
- [x] Adds type testing on PR checks

We need to be sure that the API clients response will not change without
us knowing it. As we are not the owner of the API, we can't be sure that
breaking changes will occurs silently in the future.

By introducing snapshot regression testing, we can check periodically
that the API client response is not changing.

Besides, the snapshot are used to mock the API client in E2E testing.
The means that we can test the whole application without having to mock
the API client manually.

The test interface is splitted coherently and allows for full chain test
of the application.
```
+------+
| APIs |
+------+
    |
    v
+------+
|Client|
+------+
    |
    |  <-- Snapshot testing
    v
+------------+
|Application |
+------------+
    |
    |  <-- E2E testing
    v
+------+
| User |
+------+
```

fix #643

### Fragility / flaky tests
This approach leads to create a LOT of test for non regression. A little too much maybe, given that almost everyone of them can be potentially invalidated by a change of state in real world (new company created, old one modified, etc.)

Perhaps we should extends the approach began in [clients/sirene-insee/siret/_test/index.test.ts](https://github.com/etalab/annuaire-entreprises-site/pull/651/files#diff-27963f9517072fc0a25971673d95a0cc237ccbcdc7b8397eefec76d82a54163d) by differencing code used to generate fixtures for E2E and the cases we actually test for non regression.

### How to test for failure ?

With this implementation, we do not test for failure, as we do not mock Error thrown by client. It would be interesting to have E2E test for those cases as well, as stated in #371

###  

Some clients are browser only (see #648). We should be able to stubs them for E2E test as well, in the same fashion as we test node clients.

We could do so by making `stubClient` front/back agnostic, or by creating a `stubBrowserApi` function. We could use wildcard import.

### CODE REVIEW

**Comment by XavierJp** [comment]
closes #643

File: .github/workflows/seo.yml [hunk 1]
```diff
@@ -13,7 +13,6 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
-      - uses: actions/checkout@v3
       - name: Use Node.js ${{ env.node-version }}
```
Comment (XavierJp) [inline .github/workflows/seo.yml:16]: Why did you remove the checkout here ?

**Review by XavierJp — APPROVED** [review]
A few additionnal cleaning :
- remove the old github action recipe-staging
- eventually we can remove `clients-mocks/index.ts`
- we can also clean package.json (remove "test:recipe-staging")
- maybe add a dedicated npm script to save new snapshots

File: clients/_test/clientAllEtablissementsInsee/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    '883010316',
+    '880878145',
+    '908595879',
+  ].forEach((siren) => expectClientToMatchSnapshotWithSiren(siren));
+  // expectClientToMatchSnapshotWithSiren('356000000', 1);
+  // expectClientToMatchSnapshotWithSiren('356000000', 1);
+  // expectClientToMatchSnapshotWithSiren('356000000', 3);
+  // expectClientToMatchSnapshotWithSiren('356000000', 5);
+  // expectClientToMatchSnapshotWithSiren('356000000', 6);
+  // expectClientToMatchSnapshotWithSiren('356000000', 7);
```
Comment (XavierJp) [inline clients/_test/clientAllEtablissementsInsee/index.test.ts:28]: Are these commented lines also used ?

**Review by XavierJp — APPROVED** [review]
Seems like there is a bunch of empty folders/files :
`clients/_test/clientCommuneByCp/_snapshots/.gitkeep`
`clients/_test/clientCommuneByCp/_snapshots/`
`clients/_test/clientEtablissementInsee/_snapshots/.gitkeep`

Is there a need for these empty folders ?

Really like the dedicated test folder in `/clients`

In snapshots a lot of the time args are already declare in the snapshot file why do we redeclare it ?

File: .github/workflows/seo.yml [hunk 3]
```diff
@@ -13,7 +13,8 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
-      - uses: actions/checkout@v3
+      - name: Checkout
+        uses: actions/checkout@v2
```
[the remark this answers is not in the document]
Comment (johangirod) [reply .github/workflows/seo.yml:17]: Just a bad autocomplete. Nice catch.

File: clients/_test/clientAllEtablissementsInsee/index.test.ts (hunk 2 above)
Comment (johangirod) [reply clients/_test/clientAllEtablissementsInsee/index.test.ts:28]: Yes, for the pagination cypress testing

File: setup-jest.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { toMatchFile } from 'jest-file-snapshot';
+import { disconnect } from '#utils/network/backend/redis/redis-storage';
+import 'dotenv/config';
+
+expect.extend({ toMatchFile });
+
+afterAll(async () => {
+  await disconnect();
+});
```
[the remark this answers is not in the document]
Comment (johangirod) [reply setup-jest.ts:9]: Otherwise, the handler is left hanging, and jest doesn't stop. It's not a problem locally, as you can SIGINT the process, but on the CI, the step never ends.

File: .github/workflows/pre-merge.yml [hunk 5]
```diff
@@ -161,3 +127,70 @@ jobs:
...
+        uses: ./.github/actions/load-dependencies
+      - name: Run typescript check
+        run: npx tsc --noEmit
+
+  testAPIClient:
+    name: Test API client
+    runs-on: ubuntu-latest
+    needs: [dependencies]
+    env:
+      END2END_MOCKING: disabled
```
[the remark this answers is not in the document]
Comment (johangirod) [reply .github/workflows/pre-merge.yml:151]: For the env variable, the mocking is enabled by default in the env section of the workflow. I can disable it there, but then I'll have to enable it for the build / E2E steps.

**Comment by johangirod** [comment]
> Is there a need for these empty folders ?

The actual behaviour is to throw an error if we use `stubClient({ clientName })` without creating folder `clientName` in `_test`. It ensures that we didn't make a typo that would be silently ignored.

I stubbed those clients, but they are not called in the current cypress test. We might want to create new scenarios which test them as well.

Meanwhile, I left the empty folder to be populated once it's the case.

> In snapshots a lot of the time args are already declare in the snapshot file why do we redeclare it ?

The args of the `.json` snaphots files are generated by those declared in the `*.test.ts` files. The first time you launch the test, the file is created with both the args and the response.

File: .github/workflows/pre-merge.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (XavierJp) [reply .github/workflows/pre-merge.yml:151]: Would'not it be clearer to only enbale it where it is useful ?

File: setup-jest.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (XavierJp) [reply setup-jest.ts:9]: Oh nice

File: utils/network/backend/redis/redis-storage.ts [hunk 6]
```diff
@@ -12,7 +12,11 @@ redisClient.on('error', (err) => {
   logWarningInSentry('Error in Redis', { details: err });
 });
 
-redisClient.connect();
+export async function connect() {
+  if (!redisClient.isOpen) {
+    return redisClient.connect();
+  }
+}
```
[the remark this answers is not in the document]
Comment (johangirod) [reply utils/network/backend/redis/redis-storage.ts:19]: I've tested it, it stills fails when the redis client is missing on build. See https://github.com/etalab/annuaire-entreprises-site/actions/runs/6470701786/job/17567501903?pr=651. As the client is initialized with the module, this behaviour is still unchanged. I'll add a warning if something bad happens though.

**Review by XavierJp — APPROVED** [review]
